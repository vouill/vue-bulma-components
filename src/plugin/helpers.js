/**
 * Created by Vouill on 23/07/17.
 */
/*
*   By default, components are rendered as div. This is not useful for components such as input for example.
*   This map makes possible a custom default value for the following components.
*   Key: Bulma component name ; Value: default rendered html tag
*/
export const vueBulmaDefaultRenderElement = new Map([
  ['input', 'input'],
  ['label', 'label'],
  ['button', 'button'],
  ['image', 'figure'],
  ['form', 'form'],
  ['icon', 'span'],
  ['fa', 'i']
])

/*
*   To avoid native HTML and vue-bulma-components collision (input for example),
*   we append the 'b-' prefix to the component name.
*   This is only done when using the Vue.use() way of installing vue-bulma-components
*/
export const nativeHtmlToVueBulma = new Map([
  ['input', 'b-input'],
  ['label', 'b-label'],
  ['image', 'b-image'],
  ['form', 'b-form'],
  ['button', 'b-button'],
  ['content', 'b-content'],
  ['progress', 'b-progress'],
  ['table', 'b-table'],
  ['title', 'b-title'],
  ['menu', 'b-menu'],
  ['section', 'b-section'],
  ['footer', 'b-footer']
])

/*
*   This is the list of all available vue-bulma-components rendered when using Vue.use(). If one is missing, add it here.
*/
export const bulmaComponentList = [
  'columns',
  'column',
  'notification',
  'button',
  'box',
  'icon',
  'media',
  'media-left',
  'media-right',
  'media-content',
  'image',
  'content',
  'level',
  'level-left',
  'level-item',
  'level-right',
  'field',
  'help',
  'progress',
  'table',
  'tag',
  'title',
  'subtitle',
  'breadcrumb',
  'card',
  'card-image',
  'card-content',
  'card-footer',
  'card-footer-item',
  'control',
  'label',
  'menu',
  'menu-label',
  'menu-list',
  'message',
  'message-header',
  'message-body',
  'modal',
  'modal-background',
  'modal-content',
  'modal-close',
  'navbar',
  'navbar-item',
  'navbar-brand',
  'navbar-burger',
  'navbar-start',
  'navbar-link',
  'navbar-dropdown',
  'navbar-divider',
  'pagination',
  'pagination-previous',
  'pagination-next',
  'pagination-list',
  'pagination-ellipsis',
  'panel',
  'panel-heading',
  'panel-block',
  'panel-icon',
  'panel-tabs',
  'tabs',
  'container',
  'hero',
  'hero-head',
  'hero-body',
  'input',
  'fa',
  'hero-foot'
]

export const camelCaseToDash = myStr =>
  myStr &&
  myStr
    .replace(/([a-z])([A-Z])/g, '$1-$2')
    .toLowerCase()
    .replace(/([a-z])([0-9])/g, '$1-$2')

// thanks the solution @israelroldan
export const isBulmaAttribute = attr =>
  attr.trim() && /^(is|has|fa)-.+/.test(attr)

export const getOutrEl = (outrEl, reqOutrEl, elName, defaultEl = 'div') =>
  outrEl || reqOutrEl || vueBulmaDefaultRenderElement.get(elName) || defaultEl
