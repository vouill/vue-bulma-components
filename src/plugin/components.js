/**
 * Created by Vouill on 09/07/17.
 */
import { camelCaseToDash, isBulmaAttribute, isInternalAttribute, getOutrEl } from './helpers'

export const componentGenerator = (name, reqOuterElement) => ({
  name: `bulma-${name}`,
  functional: true,
  render (h, { children, props, data }) {
    const { outerElement, ...otherProps } = props

    /*
     * Remove attribute only if it is from Bulma
     */
    if (data.hasOwnProperty('attrs')) {
      Object.keys(data.attrs).forEach(key => {
        if (isBulmaAttribute(key) || isInternalAttribute(key)) delete data.attrs[key]
      })
    }

    const bulmaClasses = Object.keys(otherProps)
      .filter(key => otherProps[key] !== false)
      .map(str => camelCaseToDash(str))
      .filter(key => isBulmaAttribute(key))

    const attrs = {
      ...data,
      class: [
        camelCaseToDash(name),
        ...bulmaClasses,
        data.class
      ]
    }

    return h(getOutrEl(outerElement, reqOuterElement, name), attrs, children)
  }
})
