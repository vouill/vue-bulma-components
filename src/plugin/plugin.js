import { componentGenerator } from './components'
import { nativeHtmlToVueBulma, bulmaComponentList } from './helpers'

let plugin = {}

plugin.install = function (Vue, option) {
  bulmaComponentList.forEach(name => {
    Vue.component(`${(option && option.prefix) || ''}${nativeHtmlToVueBulma.get(name) || name}`, componentGenerator(name))
  })
}

export default plugin
export const bulmaComponentGenerator = componentGenerator
