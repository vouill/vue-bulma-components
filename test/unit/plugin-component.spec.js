import Vue from 'vue/dist/vue.js'
import { shallow } from 'vue-test-utils'
import { componentGenerator } from 'src/plugin/components'

describe('Bulma component generator ', () => {
  it('should create a basic bulma component', () => {
    const wrapper = shallow(componentGenerator('panel'), {
      context: {}
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should trigger events', () => {
    const onClick = jest.fn()
    const wrapper = shallow(componentGenerator('box'), {
      context: { props: { isPrimary: true }, on: { click: onClick } }
    })
    expect(onClick.mock.calls.length).toBe(0)
    wrapper.trigger('click')
    expect(onClick.mock.calls.length).toBe(1)
  })
  it('should create a basic bulma component w/ children', () => {
    const vm = new Vue({
      template: '<div><box><span>children-test</span></box></div>',
      components: {
        box: componentGenerator('box')
      }
    }).$mount()
    expect(vm.$el.querySelector('span').textContent).toEqual('children-test')
  })

  it('should create a basic bulma component with modifiers', () => {
    const wrapper = shallow(componentGenerator('box', 'span'), {
      context: {
        props: { isPrimary: true }
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    const wrapper2 = shallow(componentGenerator('box'), {
      context: {
        props: { 'is-2': true }
      }
    })
    expect(wrapper2.html()).toMatchSnapshot()
  })

  it('should create a basic bulma component with modifiers binded to false', () => {
    const wrapper = shallow(componentGenerator('box'), {
      context: {
        props: { isPrimary: false }
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should create a empty div w/ no class', () => {
    const wrapper = shallow(componentGenerator(), {
      context: {}
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should create input', () => {
    const wrapper = shallow(componentGenerator('input'), {
      context: {}
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should delete added props', () => {
    const wrapper = shallow(componentGenerator('container'), {
      context: {
        props: {
          isFluid: false,
          'is-widescreen': true
        }
      }
    })

    expect(wrapper.hasProp('is-fluid')).toBe(false)
    expect(wrapper.hasProp('is-widescreen')).toBe(false)
  })
})
