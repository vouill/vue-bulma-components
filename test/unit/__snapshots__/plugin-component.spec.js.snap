// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bulma component generator  should create a basic bulma component 1`] = `"<div class=\\"panel\\"></div>"`;

exports[`Bulma component generator  should create a basic bulma component with modifiers 1`] = `"<span class=\\"box is-primary\\"></span>"`;

exports[`Bulma component generator  should create a basic bulma component with modifiers 2`] = `"<div class=\\"box is-2\\"></div>"`;

exports[`Bulma component generator  should create a basic bulma component with modifiers binded to false 1`] = `"<div class=\\"box\\"></div>"`;

exports[`Bulma component generator  should create a empty div w/ no class 1`] = `"<div class=\\"\\"></div>"`;

exports[`Bulma component generator  should create input 1`] = `"<input class=\\"input\\">"`;
